### DISCUSSION TITLE
Need to pass additional arguments to ssh ?

### DISCUSSION THREAD

**Discussion by SadarSSI** [kingToolbox/WindTerm#557] [Q&A]
I need to pass additional arguments to ssh like proxycommand.

Currently I have to go through a cmd or powershell command line to make ssh connections in certain situations: ssh tunnels with proxycommand allowing to bypass corporate or hotel proxy. In this case, I no longer have the possibility of using the sftp since the original shell is not ssh

Could you consider adding an additional field in the ssh template in order to add options / arguments to pass to the ssh command?

Some examples :

**ssh options to bypass a corporate proxy with windows authentication :**
`ssh user@server -D 10998 -o proxycmd "wstunnel.exe -vvv -L stdio:%host:%port --httpProxy=%USERDOMAIN%\%USERNAME%:%PROXYPASS%@%prxy%:%prxyport% wss: //gateway.mydomain.fr"`

**To make an ssh tunnel for an RDP session with port forwarding**
`ssh user@server -NL 2222:my-windows:3389 -o proxycmd "wstunnel.exe -vvv -L stdio:%host:%port --httpProxy=%USERDOMAIN%\%USERNAME%:%PROXYPASS%@%prxy% :%prxyport% wss://gateway.mydomain.fr`

If you could evaluate windows variables like %USERDOMAIN%\%USERNAME%:%PROXYPASS%@%prxy% :%prxyport% that would be just wonderful! (_For security reasons, I do not leave my username and password in configuration files_)

**Comment by kingToolbox** [comment]
WindTerm was designed to make it easier to use without having to memorize and type a lot of command-line arguments. Please read [Port Forwarding (Tunnel)](https://kingtoolbox.github.io/2020/07/21/port_forwarding/) and [Network Proxy / Jump Server](https://kingtoolbox.github.io/2021/03/11/proxy-jump-server/), I think they might be what you want.

If you have any further needs, don't hesitate to let me know.

BTW:
I'm so sorry that I didn't reply to your #507 in time due to an oversight.

**Comment by SadarSSI** [comment]
No, the ssh tunnel and proxy parameters do not work because the proxycommand parameter allows calling a third party program to initiate the connection like wstunnel or proxytunnel with windows or netcat[nc] with linux/mac for example.

Making an ssh tunnel using websockets is common to bypass a proxy that only lets https pass through port 443. Whether it's to ssh into remote servers, an ssh tunnel to make rdp or do reverse ssh, there are so many possibilities (...)

I think leaving a field allowing to enter optional ssh arguments could allow [geeks like me] to have the best ssh options because there would be a lot of options to put in a UI.

FYI, my putty sessions that use proxycmd parameter [alias proxycommand in ssh arguments] don't work with windterm.

So I would be very interested in having a field to manually add options to the ssh command because your program is quite fantastic and highly configurable and that's fine

**Comment by kingToolbox** [comment]
Yes, although the current design can meet the needs of most people, it cannot completely replace the functions of `netcat` and so on. I have explained the [ProxyCommand](https://kingtoolbox.github.io/2021/03/11/proxycommand/) here.

If I add a `proxy` type, including `host`, `port` and `local proxy command`. At the beginning, `WindTerm` execute the `local proxy command` first, and then send the data to `host:port`. Do you think this design will work?

**Comment by SadarSSI** [comment]
The ssh api documentation on [ https://api.libssh.org/,](url) indicates that the ssh_options_get() function [https://api.libssh.org/stable/group__libssh__session.html](url) allows to pass this type of parameters

If we consider adding a field named something like "ssh command with user options : ", it would then be implicit that all the parameters present in this field would be transmitted to this function, with beforehand an evaluation of the variables %var% for windows $var for linux.

There is some programming example on GitHub for this, unfortunately my level of programming will not be enough to help you anymore

So, yes I think this should work as the ssh_options_get() function seems to be able to handle this need

Hope that answered your question

**Reply by kingToolbox** [reply]
Thank you, actually `libssh` already supports `SSH_OPTIONS_PROXYCOMMAND`, The reason why I didn't use it is because this part of the code only supports `Linux / macOS` system, not `Windows` system. And I also didn't find a `Windows API` equivalent to the Linux `dup2` function. I couldn't get `WindTerm` to support `ProxyCommand` in `Windows`, `Linux` and `MacOS` at the same time, and dropped this feature at the time.

So if your OS is `Linux/macOS` then I'll be able to implement this soon, but if it's `Windows` then I'll need to look into `putty` to see how it's implemented. It's even possible that this won't be implemented on `Windows` as I've tried it once before.

``` c
// libssh 0.9.6 src/socket.c:873
#ifndef _WIN32
/**
 * @internal
 * @brief executes a command and redirect input and outputs
 * @param command command to execute
 * @param in input file descriptor
 * @param out output file descriptor
 */
void
ssh_execute_command(const char *command, socket_t in, socket_t out)
{
    const char *args[] = {"/bin/sh", "-c", command, NULL};
    /* Prepare /dev/null socket for the stderr redirection */
    int devnull = open("/dev/null", O_WRONLY);
    if (devnull == -1) {
        SSH_LOG(SSH_LOG_WARNING, "Failed to open /dev/null");
        exit(1);
    }

    /* redirect in and out to stdin, stdout */
    dup2(in, 0);
    dup2(out, 1);
    /* Ignore anything on the stderr */
    dup2(devnull, STDERR_FILENO);
    close(in);
    close(out);
    execv(args[0], (char * const *)args);
    exit(1);
}
...
#endif // _WIN32
```

**Comment by SadarSSI** [comment]
Generally I need this functionality WITH Windows ;-((

Previously I used the ssh port via cygwin which allowed me to do this, but since the availability of openssh 8.x on Windows (via the download on the microsoft site), I use this version.

I think the sources for the cygwin (and openssh 8.x) ssh client library can be found at [https://www.cygwin.com/cgi-bin2/package-grep.cgi?grep=openssh&arch=x86_64](url)

For the moment I go through a cmd.exe command line to launch my openssh client from winterm ;-))

Indeed, one solution would be to pass the string to an executable hosted on the host system by performing I/O redirection if I understand your code correctly, although I'm not a programming expert
Quick & dirty! :-//

**Reply by kingToolbox** [reply]
I have been reading the relevant code of putty's `PuTTY Configuration Dialog - Connection - Proxy - Telnet command, or local proxy command` for the past two days. I think this is also `ProxyCommand`, but I am not sure if this is what you want. Have you used this feature?

Both `cygwin` and `openssh` are ported from `Linux`, so reading their code doesn't help much.. For example, The minimum installation of `cygwin` is also tens of megabytes. but for `WindTerm`, it is obvious that linking so many dynamic link libraries for one feature is a bit unacceptable.

**Comment by SadarSSI** [comment]
Yes, I use this feature of putty and it works perfectly for my needs.

Yes, I agree with you, it wouldn't be a good idea to weigh down/penalize winterm with cygwin.
For information, the openssh version downloadable from the Microsoft site is 924Kb and allows the proxycommand functionality.
Searching the sources for this release, I found [https://github.com/PowerShell/openssh-portable.](url) This version is presented as compatible, linux, apple, windows (...)
there is talk about porting the proxycommand function here: [https://github.com/PowerShell/openssh-portable/tree/latestw_all/contrib/README](url) and there is explanation for compiling here [https://github.com/ powershell/openssh-portable/tree/latestw_all/contrib/win32/openssh](url)
I haven't looked at the source content to see how this feature is implemented :-//

**Reply by kingToolbox** [reply]
Sorry for the late reply. I need to take care of some stuff recently.

I have completed the `ProxyCommand` feature, you can download the latest [WindTerm_2.4.0_Prerelease_1](https://github.com/kingToolbox/WindTerm/releases/tag/2.4-prerelease) version to try it out. Due to time constraints, I did not do a complete test. Please let me know if there are any questions. Thank you.

**Reply by kingToolbox** [reply]
Sorry, I forgot to explain how to use it. After downloading, open the session's settings dialog and set `SSH - SSH - Proxy - Proxy Type` to `Porxy Command` and fill the `Command` and `Working directory`, then restart the session.

**Comment by SadarSSI** [comment]
Hello, I will test this today. Can you tell me how windows variables should be %MyVar% or ${MyVar} or $(MyVar) ?

Ok, I tried without passing variables and Ok, if in proxycommand I call a program with a windows path like c:\Tools\Common\wstunnel.exe, I get "The session is disconnected" and the log is empty.

If I put C:/Tools/Common/wstunnel.exe, I have "Unknown error" but I don't see any error in the log

An idea ?

**Comment by SadarSSI** [comment]
I just found the explanation !

If my command is "C:/Personnel/Tools/common/wstunnel.exe -vvv" (where -vvv is the verbosity level, it doesn't work). I think there is a problem with the output of the proxycommand command. It is played before the ssh session, is that why it is not visible in the log even in debug mode?

If I remove -vvv it's work ! ;-))

Last note : I failed to get a windows variable evaluated in the proxycommand line by trying these combinations : %MyVar% (windows mode), ${MyVar} or $(MyVar). A suggestion ?

In any case thank you for this version which improves this part

**Comment by kingToolbox** [comment]
If you set the `ProxyCommand` to `C:/Personnel/Tools/common/wstunnel.exe`, `WindTerm` will read all the outputs of `wstunnel.exe`, and write all the inputs to `wstunnel.exe`. So I'm not quite sure if adding `-vvv` will make `wstunnel.exe` have extra output and cause it to fail, I will analyze this further.

In `Windows`, if you need to use a system environment variable in the command line, then you may write it as `cmd /c C:/Personnel/Tools/common/wstunnel.exe %env_variable%`, otherwise the environment variable will not take effect. But I have implemented the replacement of system environment variables in the command line with its value, so in the next `WindTerm_2.4.0_Prerelease_2` version you just need to write it as `C:/Personnel/Tools/common/wstunnel.exe %env_variable%`, which is more in line with the usage habit, and more importantly, this is cross-platform.

Also I added a `custom environment variables listbox`, just like `Session Settings - Shell - Process - Environment variables`.

The development of `WindTerm`'s own variable feature has not yet been completed, and only `$(HomeDir)` and `$(AppDir)` can be parsed at present.

**Comment by SadarSSI** [comment]
Sorry for my late reply too

For wstunnel verbosity, look at what happens for terminal redirection as it happens BEFORE the ssh connection. I don't know how WINDTERM reacted in this specific case.

Thanks for the implementation tips for the variables, I'll wait for the next pre-release to test this.

Tunnel operation mode, not working properly

- the local tunnel stops after a few seconds / minutes
- the dynamic tunnel does not work (netstat -an see the open port, but impossible to connect to it)

I hope that the libssh will allow the Windows version to evolve with the same functionalities that we have with the openssh version available in command line [https://github.com/PowerShell/openssh-portable](https://github.com/PowerShell/openssh-portable) like I told you in our previous conversation

**Reply by kingToolbox** [reply]
At present, the `Windows` version of `ProxyCommand` is completely implemented by me, and has nothing to do with `libssh`. `libssh` only provides the `Linux` version of `ProxyCommand`. Since I've been working hard on the `Free Type Mode` and `Dynamic Memory Compression` features, I haven't had time to read the `openssh-portable` code, but I'll look into it after the new version is released to see if there's anything I can learn from. The new `WindTerm_2.4.0_Prerelease_2` version will be released in two or three days.

Thank you for pointing out the problem with `tunnels` too. I'll leave the these issues to be fixed in `WindTerm_2.4.0_Prerelease_3`.

**Reply by kingToolbox** [reply]
I have done and released the [WindTerm_2.4.0_Prerelease_2](https://github.com/kingToolbox/WindTerm/releases/tag/2.4-prerelease) version. Now you can customize environment variables. Please try it out and see if it meets your needs.

The two problems with `Tunnels` you mentioned, the first one may be closed by the server due to the idle timeout on the server side. I have fixed it by adding a `keepAlive` mechanism. The second question I haven't checked, but I've added more error messages, so if there are still problems, maybe that can help.

**Comment by SadarSSI** [comment]
I just tested this new prerealse 2 version

I tried declaring the windows variables in the box list and couldn't get them to work with the proxycommand line.
For example, I have a prxy variable declared at user level. I can declare it with the listbox, but when I declare it in the command line it doesn't work. I've tried various syntaxes for this: $(prxy), $prxy, or %prxy% , but it doesn't work
Regarding the declaration of variables, will the value associated with this variable be evaluated each time or will it be the initially declared value that will be taken?
Indeed, my variables; user-side variables, such as those declared in a .bashrc file under linux, change value depending on certain context or needs. Thus the variable $(prxy): %proxy% under windows cmd or $env:prxy in powershell, can change value depending on the proxy I want to use. In reality I have several variables allowing me to address different proxies depending on my location and the job I am doing. So in my daily use, my proxycommand string is dynamically evaluated based on variables and their contents.
At the moment I continue to go through a command line via a terminal with WINDTERM with a cmd session, powershell session or cygwin because I can't use WSL at my work ;-((, and as you probably know, the windows variables are simply accessible under cygwin with a bash-like syntax like $prx. What's a shame about doing this is that we can't benefit from sftp on the target host ;-//

wstunnel's -vv option always causes the same error

For the tunnel functionality, when I try to activate the tunnel on an existing session (with proxycommand) I have the message "waiting to open" which never ends and it does not work.

Personally, I think the idea that passing the additional parameters for the ssh command in a dedicated field would be easier for the user experience. Indeed, it is quite confusing to go to a menu to activate a tunnel option and, in my case, I would have to declare several tunnels to manage to do both a dynamic port and port forwarding for the same session.

In any case, I now use WINDTERM every day in personal and professional environments.
